apps/emqx_resource/src/emqx_resource_worker.erl

```diff
@@ -252,145 +274,274 @@ code_change(_OldVsn, State, _Extra) ->
 ).
 
 pick_call(Id, Key, Query, Timeout) ->
-    ?PICK(Id, Key, gen_statem:call(Pid, Query, {clean_timeout, Timeout})).
+    ?PICK(Id, Key, begin
+        Caller = self(),
+        MRef = erlang:monitor(process, Pid, [{alias, reply_demonitor}]),
+        From = {Caller, MRef},
+        erlang:send(Pid, ?SEND_REQ(From, Query)),
+        receive
+            {MRef, Response} ->
+                erlang:demonitor(MRef, [flush]),
+                Response;
+            {'DOWN', MRef, process, Pid, Reason} ->
+                error({worker_down, Reason})
+        after Timeout ->
+            erlang:demonitor(MRef, [flush]),
+            receive
+                {MRef, Response} ->
+                    Response
+            after 0 ->
+                error(timeout)
+            end
+        end
+    end).
 
 pick_cast(Id, Key, Query) ->
-    ?PICK(Id, Key, gen_statem:cast(Pid, Query)).
+    ?PICK(Id, Key, begin
+        From = undefined,
+        erlang:send(Pid, ?SEND_REQ(From, Query)),
+        ok
+    end).
 
-do_resume(#{id := Id, name := Name} = St) ->
+do_resume(#{id := Id, name := Name} = Data) ->
     case inflight_get_first(Name) of
         empty ->
-            retry_queue(St);
+            retry_queue(Data);
         {Ref, FirstQuery} ->
             %% We retry msgs in inflight window sync, as if we send them
             %% async, they will be appended to the end of inflight window again.
-            retry_inflight_sync(Id, Ref, FirstQuery, Name, St)
+            retry_inflight_sync(Id, Ref, FirstQuery, Name, Data)
     end.
 
-retry_queue(#{queue := undefined} = St) ->
-    {next_state, running, St};
 retry_queue(
     #{
-        queue := Q,
+        queue := Q0,
         id := Id,
         index := Index,
-        enable_batch := false,
+        batch_size := 1,
+        name := Name,
         resume_interval := ResumeT
-    } = St
+    } = Data0
 ) ->
-    case get_first_n_from_queue(Q, 1) of
-        [] ->
-            {next_state, running, St};
-        [?QUERY(_, Request, HasSent) = Query] ->
-            QueryOpts = #{inflight_name => maps:get(name, St)},
+    %% no batching
+    case get_first_n_from_queue(Q0, 1) of
+        empty ->
+            {next_state, running, Data0};
+        {Q1, QAckRef, [?QUERY(_, Request, HasBeenSent) = Query]} ->
+            QueryOpts = #{inflight_name => Name},
             Result = call_query(configured, Id, Index, Query, QueryOpts),
-            case reply_caller(Id, ?REPLY(undefined, Request, HasSent, Result)) of
+            Reply = ?REPLY(undefined, Request, HasBeenSent, Result),
+            case reply_caller(Id, Reply) of
                 true ->
-                    {keep_state, St, {state_timeout, ResumeT, resume}};
+                    {keep_state, Data0, {state_timeout, ResumeT, resume}};
```

Comment:
Before this refactoring, the only requests that did bump the `retried.*` counters were the async requests, and they are the only ones that do get inserted into the inflight table and marked as sent.

I'm not entirely sure this was the intended behavior indeed.  But I think that this behavior is kept because, after this refactoring, the path that an async request does is:

1. The request gets enqueued into replayq.
2. It's then popped (possible in a batch).
3. It's appended to the inflight table (if there's room).
4. In the first ("outer") `handle_query_result`, since it'll return `{async_return, ok}`, it'll be ack'ed in replayq, removing it from there.
5. Eventually, when `{batch_,}reply_after_query` is called, then it'll bump either `retried.*` counter.

For sync requests, you are right: they won't bump the `retried.*` counters.  Do you think that's the original intention?  We would need to track what has been retried somehow.  

We can't change the replayq items without re-appending them to the queue, and that would change the order of the requests, unfortunately.  One way could be to keep a table of hashes of requests and then check if those have been sent or not.  :thinking: 

What do you think?

Reply:
Personally, I think we can just ignore this here, then create a track ticket to the product owner, and left the discussion to them

Reply:
Sounds good to me: we can revisit this point in a future PR, if required.  :smiley_cat: